src/box/lua/config/applier/roles.lua

```diff
@@ -1,16 +1,78 @@
 local log = require('internal.config.utils.log')
+local fiber = require('fiber')
 
-local last_loaded = {}
-local last_roles_ordered = {}
+local roles_state = {
+    -- Loaded roles, where the key is the role name and the value is the
+    -- role object.
+    last_loaded = {},
+
+    -- Ordered list of roles, where the order is determined by the
+    -- dependencies between roles.
+    last_roles_ordered = {},
+
+    -- The box.status watcher is used to call on_box_status callbacks.
+    box_status_watcher = nil,
+
+    -- Last values received from box.status watcher.
+    last_box_status_value = nil,
+
+    -- Last applied config.
+    last_config = nil,
+}
+
+-- Call all the 'on_box_status' callbacks from roles in order.
+local function call_on_box_status_callbacks(config, value)
```

Comment:
I would pass an event key here and pass it to `on_box_status` (or `on_event` if we rename it). I would pass a different key for the reconfiguration event. It gives a bit more information for a user and costs us nothing.

(Maybe it also helps in testing -- we can differentiate an origin of the call.)

Reply:
Done. `key = 'box.status'` if the callback was triggered by `box.status` system event, and `key = 'config.apply'` if it was triggered because of configuration update.